src/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.PowMod.cs

```diff
@@ -0,0 +1,536 @@
+        {
+            Debug.Assert(value != null);
+
+            // The 32-bit modulus pow method for a big integer
+            // raised by a 32-bit integer...
+
+            uint temp = Remainder(value, modulus)[0];
+
+            uint result = PowCore(power, modulus, temp, 1);
+
+            return new uint[] { result };
+        }
+
+        public static uint[] Pow(uint value, uint[] power, uint modulus)
+        {
+            Debug.Assert(power != null);
+
+            // The 32-bit modulus pow method for a 32-bit integer
+            // raised by a big integer...
+
+            uint result = PowCore(power, modulus, value, 1);
+
+            return new uint[] { result };
+        }
+
+        public static uint[] Pow(uint[] value, uint[] power, uint modulus)
+        {
+            Debug.Assert(value != null);
+            Debug.Assert(power != null);
+
+            // The 32-bit modulus pow method for a big integer
+            // raised by a big integer...
+
+            uint temp = Remainder(value, modulus)[0];
+
+            uint result = PowCore(power, modulus, temp, 1);
+
+            return new uint[] { result };
+        }
+
+        private static uint PowCore(uint[] power, uint modulus,
+                                    ulong temp, ulong result)
+        {
+            // The 32-bit modulus pow algorithm for all but
+            // the last power limb using square-and-multiply.
+
+            for (int i = 0; i < power.Length - 1; i++)
+            {
+                uint p = power[i];
+                for (int j = 0; j < 32; j++)
+                {
+                    if ((p & 1) == 1)
+                        result = (result * temp) % modulus;
+                    temp = (temp * temp) % modulus;
+                    p = p >> 1;
+                }
+            }
+
+            return PowCore(power[power.Length - 1], modulus, temp, result);
+        }
+
+        private static uint PowCore(uint power, uint modulus,
+                                    ulong temp, ulong result)
+        {
+            // The 32-bit modulus pow algorithm for the last or
+            // the only power limb using square-and-multiply.
+
+            while (power != 0)
+            {
+                if ((power & 1) == 1)
+                    result = (result * temp) % modulus;
+                temp = (temp * temp) % modulus;
+                power = power >> 1;
+            }
+
+            return (uint)(result % modulus);
+        }
+
+        public static uint[] Pow(uint value, uint power, uint[] modulus)
+        {
+            Debug.Assert(modulus != null);
+
+            // The big modulus pow method for a 32-bit integer
+            // raised by a 32-bit integer...
+
+            BitsBuffer temp = new BitsBuffer(modulus.Length * 2, value);
+            BitsBuffer result = new BitsBuffer(modulus.Length * 2, 1);
+            FastReducer reducer = new FastReducer(modulus);
```

Comment:
Actually, I invested a bit more in reducing allocations, since I've learned from our previous work that these are damn crucial (my original implementation just used the big-integer structure, thus relocated with every operation...). But let's have a look at the numbers:

- The current implementation needs 4 "builders" (`regRes`, `regVal`, `regMod`, `regTmp`), the algorithm needs its values. I don't know any way how to use less than those 2 buffers.
- One "buffer" allocates 2 arrays, one for the current value and one for the next, and switches between them. So the 2 is a constant value, no matter how many calculations are needed (the bigger the numbers, the more calculations). For 32-bit integers we are even buffer-free.
- The "reducer" creates some overhead though, but the Barrett reduction needs it's μ, q and ˜q. As with the buffer the 3 is at least constant.

So, how can we squeeze this further?

- As suggested, turning buffer and reducer into structures shouldn't be any problem. I've no idea why I've chosen classes anyway.
- As already suspected at the beginning, the Barrett reduction may be not that efficient for smaller numbers (I'll investigate this further!). If that's the case, we can define a threshold for using an ordinary remainder, so we have this overhead only for the really big ones.

Sounds ok?